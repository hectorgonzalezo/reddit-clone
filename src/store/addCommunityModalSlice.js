import { createSlice } from '@reduxjs/toolkit';

export const addCommunityModalSlice = createSlice({
  name: 'addCommunityModalVisible',
  initialState: false,
  reducers: {
    toggleAddCommunityModal: (state, action) => {
      return !state
    }
  },
});

export const { toggleAddCommunityModal } = addCommunityModalSlice.actions;

export const selectAddCommunityModalVisibility = (state) => state.addCommunityModalVisible;

export default addCommunityModalSlice.reducer;
