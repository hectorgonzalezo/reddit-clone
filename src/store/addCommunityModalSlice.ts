import { createSlice } from '@reduxjs/toolkit';

export const addCommunityModalSlice = createSlice({
  name: 'addCommunityModalVisible',
  initialState: false,
  reducers: {
    toggleAddCommunityModal: (state) => !state
  },
});

export const { toggleAddCommunityModal } = addCommunityModalSlice.actions;

export const selectAddCommunityModalVisibility = (state: AddCommunityModalState): boolean => state.addCommunityModalVisible;

export default addCommunityModalSlice.reducer;
