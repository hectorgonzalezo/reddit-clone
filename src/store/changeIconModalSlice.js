import { createSlice } from '@reduxjs/toolkit';

export const changeIconModalSlice = createSlice({
  name: 'changeIconModalVisible',
  initialState: false,
  reducers: {
    toggleChangeIconModal: (state, action) => {
      return !state
    }
  },
});

export const { toggleChangeIconModal } = changeIconModalSlice.actions;

export const selectChangeIconModalVisibility = (state) => state.changeIconModalVisible;

export default changeIconModalSlice.reducer;
