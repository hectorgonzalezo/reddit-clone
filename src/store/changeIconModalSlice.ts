import { createSlice } from '@reduxjs/toolkit';
import { ChangeIconModalState } from '../type';

export const changeIconModalSlice = createSlice({
  name: 'changeIconModalVisible',
  initialState: false,
  reducers: {
    toggleChangeIconModal: (state, action) => !state
  },
});

export const { toggleChangeIconModal } = changeIconModalSlice.actions;

export const selectChangeIconModalVisibility = (state: ChangeIconModalState): boolean => state.changeIconModalVisible;

export default changeIconModalSlice.reducer;
