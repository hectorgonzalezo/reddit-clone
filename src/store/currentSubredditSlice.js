import { createSlice } from '@reduxjs/toolkit';

export const currentSubredditSlice = createSlice({
  name: 'currentSubreddit',
  initialState: null,
  reducers: {
    changeCurrentSubreddit: (state, action) => {
      return action.payload
    }
  },
});

export const { changeCurrentSubreddit } = currentSubredditSlice.actions;

export const selectCurrentSubreddit = (state) => state.currentSubreddit;

export default currentSubredditSlice.reducer;
