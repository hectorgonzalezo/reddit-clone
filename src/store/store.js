import { configureStore } from '@reduxjs/toolkit';
import userReducer from './userSlice';
import subredditsReducer from './subredditsSlice';
import loginModalReducer from './loginModalSlice';
import currentSubredditReducer from './currentSubredditSlice';

const store = configureStore({
  reducer: {
    user: userReducer,
    subreddits: subredditsReducer,
    loginModalVisible: loginModalReducer,
    currentSubreddit: currentSubredditReducer,
  }
});

export default store;
