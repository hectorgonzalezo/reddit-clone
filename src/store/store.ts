import { configureStore } from '@reduxjs/toolkit';
import userReducer from './userSlice';
import postsReducer from './postsSlice';
import subredditsReducer from './subredditsSlice';
import loginModalReducer from './loginModalSlice';
import changeIconModalReducer from './changeIconModalSlice';
import addCommunityModalReducer from './addCommunityModalSlice';
import currentSubredditReducer from './currentSubredditSlice';

const store = configureStore({
  reducer: {
    user: userReducer,
    posts: postsReducer,
    subreddits: subredditsReducer,
    loginModalVisible: loginModalReducer,
    changeIconModalVisible: changeIconModalReducer,
    addCommunityModalVisible: addCommunityModalReducer,
    currentSubreddit: currentSubredditReducer,
  }
});

export default store;
