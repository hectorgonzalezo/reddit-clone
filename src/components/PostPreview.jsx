import React, { useState } from 'react';
import { formatDistanceToNow } from 'date-fns';
import { useSelector } from 'react-redux';
import { string, number, arrayOf, object } from 'prop-types';
import { database, authorization } from '../firebase/firebase';
import styled from 'styled-components';
import upIcon from '../assets/upvote_icon.svg';
import downIcon from '../assets/downvote_icon.svg';
import commentsIcon from '../assets/comments_icon.svg';
import shareIcon from '../assets/share_icon.svg';
import saveIcon from '../assets/save_icon.svg';
import hideIcon from '../assets/hide_icon.svg';
import IconLink from './IconLink';
import SubredditIcon from './SubredditIcon';
import formatUpVotes from '../utils/formatUpVotes';
import { selectUser } from '../store/userSlice';

const Post = styled.article`
  display: grid;
  gap: 10px;
  padding-right: 0px;
  grid-template-columns: 40px 1fr;
  grid-template-rows: 20px 1fr 30px;
  grid-template-areas:
    "vote sub"
    "vote main"
    "vote comm";

  & > div {
    padding-right: 10px;
  }

  em {
    font-weight: bold;
    color: black;
  }

  div {
    display: flex;
  }
  &:hover {
    outline: 1px solid grey;
    cursor: pointer;
  }

  & > .vote-area-post {
    grid-area: vote;
    flex-direction: column;
    align-items: center;
    justify-content: flex-start;
    gap: 5px;
    img {
      width: 20px;
    }
    p {
      font-weight: bold;
      color: ${(props) =>
        props.voteType === "upVote"
          ? "var(--up-vote-color)"
          : props.voteType === "downVote"
          ? "var(--reddit-blue-light)"
          : "black"};
    }
  }

  & > .top-area-post {
    grid-area: sub;
    gap: 3px;
    a:hover {
      text-decoration: underline;
    }
    a,
    p {
      font-size: 0.8rem;
      color: grey;
      display: flex;
      align-items: center;
      gap: 3px;
    }
  }

  & > .main-area-post {
    grid-area: main;
    display: flex;
    flex-direction: column;
    gap: 15px;
    padding-right: 0px;
    & > *:not(img) {
      padding-right: 13px;
    }
    img {
      max-width: 100%;
    }
    p {
      font-size: 0.82rem;
      line-height: 1.5;
      white-space: pre-wrap;
      max-height: 250px;
      overflow: hidden;
      -webkit-mask-image: -webkit-gradient(
        linear,
        left 70%,
        left bottom,
        from(rgba(0, 0, 0, 1)),
        to(rgba(0, 0, 0, 0))
      );
    }
  }

  & > .bottom-area-post {
    grid-area: comm;
  }
`;

function PostPreview({
  subredditName,
  subredditIcon,
  postId,
  voteType,
  poster,
  title,
  text,
  upVotes,
  timePosted,
  comments,
  img,
}) {
  const [previousVote, setPreviousVote] = useState(voteType);
  const [votes, setVotes] = useState(upVotes);
  const user = useSelector(selectUser);
  // updates number of upvotes in post
  function updateVotes(e) {
    // Only allow vote if user is authorized
    if (authorization.getUser() !== null) {
      // if(authorization.user)
      // can be 'upVote' or 'downVote'
      const newVoteType = e.target.getAttribute('data');
      switch (true) {
        // previously clicked vote => new vote
        // upVote => upVote
        case newVoteType === 'upVote' && previousVote === 'upVote':
          setPreviousVote('');
          setVotes((prevVotes) => prevVotes - 1);
          database.updateVotes(user.username, subredditName, postId, -1, '');
          break;
        // none => upVote
        case newVoteType === 'upVote' && previousVote === '':
          setPreviousVote('upVote');
          setVotes((prevVotes) => prevVotes + 1);
          database.updateVotes(
            user.username,
            subredditName,
            postId,
            1,
            newVoteType
          );
          break;
        // downVote => downVote
        case newVoteType === "downVote" && previousVote === "downVote":
          setPreviousVote("");
          setVotes((prevVotes) => prevVotes + 1);
          database.updateVotes(user.username, subredditName, postId, 1, "");
          break;
        // none => downVote
        case newVoteType === "downVote" && previousVote === "":
          setPreviousVote("downVote");
          setVotes((prevVotes) => prevVotes - 1);
          database.updateVotes(
            user.username,
            subredditName,
            postId,
            -1,
            newVoteType
          );
          break;
        // downvote => upvote
        case newVoteType === "downVote" && previousVote === "upVote":
          setPreviousVote("downVote");
          setVotes((prevVotes) => prevVotes - 2);
          database.updateVotes(
            user.username,
            subredditName,
            postId,
            -2,
            newVoteType
          );
          break;
        // upvote => downvote
        case newVoteType === "upVote" && previousVote === "downVote":
          setPreviousVote("upVote");
          setVotes((prevVotes) => prevVotes + 2);
          database.updateVotes(
            user.username,
            subredditName,
            postId,
            2,
            newVoteType
          );
          break;
        default:
          break;
      }
    } else {
      openLogInModal();
    }
  }

  return (
    <Post className="main-child" voteType={previousVote}>
      <div className="vote-area-post">
        <IconLink fill="orange" onClick={updateVotes} data="upVote" colored={previousVote === 'upVote'}>
          <img src={upIcon} alt="" data="upVote"/>
        </IconLink>
        <p>{formatUpVotes(votes)}</p>
        <IconLink fill="blue" onClick={updateVotes} data="downVote" colored={previousVote === 'downVote'}>
          <img src={downIcon} alt="" data="downVote"/>
        </IconLink>
      </div>

      <div className="top-area-post">
        <a>
          <SubredditIcon icon={subredditIcon} small />
          <em>r/{subredditName}</em>
        </a>
        <p>&nbsp;â€¢&nbsp;</p>
        <p>
          {" "}
          Posted by <a href="">u/{poster}</a>{" "}
          {formatDistanceToNow(new Date(timePosted))} ago
        </p>
      </div>

      <div className="main-area-post">
        <h1>{title}</h1>
        {/* display image if any is provided */}
        {img !== '' ? <img src={img} alt="Content image" data-testid="image-content" /> : null}
        {/* display text if any is provided */}
        {text !== '' ? <p>{text}</p> : null}
      </div>

      <div className="bottom-area-post">
        <IconLink>
          <img src={commentsIcon} alt="" className="icon" />
          <p>{formatUpVotes(comments.length)} Comments</p>
        </IconLink>
        <IconLink>
          <img src={shareIcon} alt="" className="icon" />
          <p>Share</p>
        </IconLink>
        <IconLink>
          <img src={saveIcon} alt="" className="icon" />
          <p>Save</p>
        </IconLink>
        <IconLink>
          <img src={hideIcon} alt="" className="icon" />
          <p>Hide</p>
        </IconLink>
      </div>
    </Post>
  );
}

PostPreview.defaultProps = {
  subredditIcon: 'https://firebasestorage.googleapis.com/v0/b/reddit-clone-83ce9.appspot.com/o/default_icon.svg?alt=media&token=4b92a9a0-3b37-4058-bdca-627d706dd7d6',
  text: '',
  voteType: '',
  comments: [],
  img: '',
};

PostPreview.propTypes = {
  subredditName: string.isRequired,
  subredditIcon: string,
  postId: string.isRequired,
  poster: string.isRequired,
  title: string.isRequired,
  text: string,
  upVotes: number.isRequired,
  voteType: string,
  timePosted: string.isRequired,
  comments: arrayOf(object),
  img: string,
};

export default PostPreview;
