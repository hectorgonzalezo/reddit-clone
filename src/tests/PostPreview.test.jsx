import React from 'react';
import { render, screen, container } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import PostPreview from '../components/PostPreview';
import { Provider } from 'react-redux';
import store from '../store/store';

describe('Post previews', () => {
  const mockPost = {
    subredditName: 'aww',
    subredditIcon: undefined,
    poster: 'yo',
    title: 'Title',
    text: 'Hey!',
    imageUrl: 'https://firebasestorage.googleapis.com/v0/b/reddit-clone-83ce9.appspot.com/o/postImages%2Fbar%2Fimages58b82c35dcf6e61433da9494.png?alt=media&token=c77db72a-07d3-4c61-9a2c-cfc86e8e4bb1',
    postId: '0',
    upVotes: 2300,
    timePosted: (new Date()).toString(),
    comments: [{}],
  };

  test('Render title and basic info', () => {
    // get values from mockPost
    const {
      subredditName,
      subredditIcon,
      poster,
      title,
      upVotes,
      timePosted,
      comments,
      postId,
    } = mockPost;

    // render mock post preview
    render(
    <Provider store={store}>
      <PostPreview
        key={title}
        subredditName={subredditName}
        subredditIcon={subredditIcon}
        poster={poster}
        timePosted={timePosted}
        title={title}
        postId={postId}
        upVotes={upVotes}
        comments={comments}
      />
    </Provider>);

    expect(screen.getByRole('heading', { name: title })).toBeInTheDocument();

    const paragraphs = screen.queryByRole('paragraph');
    // console.log(paragraphs)
    expect(screen.getByRole('link', { name: `u/${poster}` })).toBeInTheDocument();
    expect(screen.getByText('2.3k')).toBeInTheDocument();
    expect(screen.getByText(`r/${subredditName}`)).toBeInTheDocument();
    expect(screen.getByText(`u/${poster}`)).toBeInTheDocument();
    expect(screen.getByText('1 Comments')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test('Render text if given one', () => {
    // get values from mockPost
    const {
      subredditName,
      subredditIcon,
      poster,
      title,
      text,
      upVotes,
      timePosted,
      comments,
      postId,
    } = mockPost;

    // render mock post preview
    render(
      <Provider store={store}>
        <PostPreview
          key={title}
          subredditName={subredditName}
          subredditIcon={subredditIcon}
          poster={poster}
          timePosted={timePosted}
          title={title}
          text={text}
          postId={postId}
          upVotes={upVotes}
          comments={comments}
        />
      </Provider>
    );

    expect(screen.getByRole('heading', { name: title })).toBeInTheDocument();

    const textContent = screen.getByText('Hey!');
    expect(textContent).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test('Given an imageUrl display it', () => {
    // get values from mockPost
    const {
      subredditName,
      subredditIcon,
      poster,
      title,
      imageUrl,
      upVotes,
      postId,
      timePosted,
      comments,
    } = mockPost;

    // render mock post preview
    render(
      <Provider store={store}>
        <PostPreview
          key={title}
          subredditName={subredditName}
          subredditIcon={subredditIcon}
          poster={poster}
          timePosted={timePosted}
          title={title}
          img={imageUrl}
          postId={postId}
          upVotes={upVotes}
          comments={comments}
        />
      </Provider>
    );

    expect(screen.getByRole('heading', { name: title })).toBeInTheDocument();

    const imageContent = screen.getByTestId('image-content');
    expect(imageContent).toBeInTheDocument();
    expect(imageContent).toHaveAttribute('src', mockPost.imageUrl);

    expect(container).toMatchSnapshot();
  });

  test.skip('Pressing upvote and downvote buttons disabled if no user is logged in', () => {
  });
  test.skip('Pressing upvote updates number and changes icon color', () => {
  });
  test.skip('Pressing downvote updates number and changes icon color', () => {
  });
  test.skip('Pressing upvote twice removes previous vote', () => {
  });
  test.skip('Pressing downvote twice removes previous vote', () => {
  });
  test.skip('Pressing upvote and then downvote leaves only downvote', () => {
  });
})