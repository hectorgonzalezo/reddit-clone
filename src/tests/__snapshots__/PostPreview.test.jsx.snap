// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post previews Given an imageUrl display it 1`] = `undefined`;

exports[`Post previews Render text if given one 1`] = `undefined`;

exports[`Post previews Render title and basic info 1`] = `undefined`;
